types/tipset.go

```diff
@@ -2,155 +2,148 @@ package types
+	return len(ts.blocks) > 0
+}
 
-	return r
+// Len returns the number of blocks in the tipset.
+func (ts TipSet) Len() int {
+	return len(ts.blocks)
 }
 
-// String returns a formatted string of the TipSet:
-// { <cid1> <cid2> <cid3> }
-func (ts TipSet) String() string {
-	return ts.ToSortedCidSet().String()
+// At returns the i'th block in the tipset.
+// An index outside the half-open range [0, Len()) will panic.
+func (ts TipSet) At(i int) *Block {
+	return ts.blocks[i]
 }
 
-// Equals returns true if the tipset contains the same blocks as another set.
-// Equality is not tested deeply.  If blocks of two tipsets are stored at
-// different memory addresses but have the same cids the tipsets will be equal.
-func (ts TipSet) Equals(ts2 TipSet) bool {
-	return ts.ToSortedCidSet().Equals(ts2.ToSortedCidSet())
+// IsSolo tests whether the tipset has a single block.
+func (ts TipSet) IsSolo() bool {
+	return len(ts.blocks) == 1
 }
 
-// ToSortedCidSet returns a SortedCidSet containing the Cids in the
-// TipSet.
+// ToSortedCidSet returns a SortedCidSet containing the CIDs in the tipset.
 func (ts TipSet) ToSortedCidSet() SortedCidSet {
 	s := SortedCidSet{}
-	for _, b := range ts {
+	for _, b := range ts.blocks {
 		s.Add(b.Cid())
 	}
 	return s
 }
 
-// ToSlice returns the slice of *Block containing the tipset's blocks.
-// Sorted.
+// ToSlice returns an ordered slice of pointers to the tipset's blocks.
 func (ts TipSet) ToSlice() []*Block {
-	sl := make([]*Block, len(ts))
-	var i int
-	for it := ts.ToSortedCidSet().Iter(); !it.Complete(); it.Next() {
-		sl[i] = ts[it.Value()]
-		i++
-	}
-	return sl
+	slice := make([]*Block, len(ts.blocks))
+	copy(slice, ts.blocks)
+	return slice
 }
 
-// MinTicket returns the smallest ticket of all blocks in the tipset.
+// MinTicket returns the smallest ticket of all blocks in the tipset, and nil error.
 func (ts TipSet) MinTicket() (Signature, error) {
-	if len(ts) == 0 {
-		return nil, ErrEmptyTipSet
-	}
-	blks := ts.ToSlice()
-	min := blks[0].Ticket
-	for i := range blks[0:] {
-		if bytes.Compare(blks[i].Ticket, min) < 0 {
-			min = blks[i].Ticket
-		}
-	}
-	return min, nil
+	return ts.blocks[0].Ticket, nil
 }
 
-// Height returns the height of a tipset.
+// Height returns the height of a tipset, and nil error.
 func (ts TipSet) Height() (uint64, error) {
-	if len(ts) == 0 {
-		return uint64(0), ErrEmptyTipSet
-	}
-	return uint64(ts.ToSlice()[0].Height), nil
+	return uint64(ts.blocks[0].Height), nil
 }
 
-// Parents returns the parents of a tipset.
+// Parents returns the CIDs of the parents of the blocks in the tipset, and nil error.
 func (ts TipSet) Parents() (SortedCidSet, error) {
-	if len(ts) == 0 {
-		return SortedCidSet{}, ErrEmptyTipSet
-	}
-	return ts.ToSlice()[0].Parents, nil
+	return ts.blocks[0].Parents, nil
 }
 
-// ParentWeight returns the tipset's ParentWeight in fixed point form.
+// ParentWeight returns the tipset's ParentWeight in fixed point form, and nil error.
```

Comment:
Also if we do decide to go with this then there should be some comment explicitly saying that these errors are intended to go away soon.  As it is this code is the worst of both worlds: we are forced to write error handling code but that error handling code never gets used because this code panics instead of passing along the error

Reply:
If this lands, I'll fast follow with removing the errors. Pending broader advice on the pattern.